packages/fx-core/src/common/azure-hosting/azureHosting.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Inputs, ResourceTemplate, Void } from "@microsoft/teamsfx-api";
+import { Context } from "@microsoft/teamsfx-api/build/v2";
+import * as fs from "fs-extra";
+import path from "path";
+import { generateBicepFromFile } from "..";
+import { ArmTemplateResult } from "../armInterface";
+import { Bicep } from "../constants";
+import { getTemplatesFolder } from "../../folder";
+import { getActivatedV2ResourcePlugins } from "../../plugins/solution/fx-solution/ResourcePluginContainer";
+import { NamedArmResourcePluginAdaptor } from "../../plugins/solution/fx-solution/v2/adaptor";
+import { BicepConfigs } from "./interface";
+
+export abstract class AzureHosting {
+  abstract hostType: string;
+  abstract configurable: boolean;
+
+  reference: any = undefined;
```

Comment:
Just curious on why reference is not abstract field as hostType/configurable?

Reply:
An abstract field means every child class needs to implement it, which cause some hosting has to have reference = undefined. 
For example:
```
class BotServiceHosting {
   hostType = "botservice";
   reference = undefined;
}
```